Add custom derivative to DerivativeSet

vaithak opened the pull request:
This is required if the definition of the custom derivative is not found in the current translation unit and is linked in from another. 
Adding it to the set of derivatives ensures that the custom derivative is not differentiated again using numerical differentiation due to an unavailable definition.

vaithak commented:
To test this locally with Clang, we can declare a custom derivative (for a called function) but cannot offer its definition. Now, computing the hessian should result in a linker error instead of a compile-time error.
A similar scenario was observed in the cling and ROOT setup, where the definition of custom derivative was not found in some cases.

Minimal code:
```cpp
#include "clad/Differentiator/Differentiator.h"

namespace clad{
namespace custom_derivatives {
  clad::ValueAndPushforward<double, double> f_pushforward(double x, double _d_x);
}
}

double f(double x) {
  return x*x;
}

double wrapper(double x) {
  return f(x);
}

int main() {
  clad::hessian(wrapper);
  return 0;
}
```

vgvassilev commented:
Can we add this test to the PR?

vaithak commented:
As mentioned above, the crude way to test this locally was seeing a linker error instead of a compile time error because of numerical differentiation. I don't think it would be good to have something like that in the CI check.

We can test this with ROOT infrastructure though.

parth-07 commented:
@vaithak We can test this by adding multiple files for the test, right? For example, we can have two files, `CustomDerivativesDecl.C` and `CustomDerivativesDef.cpp`. `CustomDerivativesDecl.C` will have declarations of custom derivatives and `CustomDerivativesDef.cpp` will have the definitions of custom derivatives. `CustomDerivativesDecl.C` will contain the actual test as well and it will utilize `CustomDerivativesDef.cpp` in the compile commands of the test.

vgvassilev commented on lib/Differentiator/DerivativeBuilder.cpp:
```diff
@@ -253,6 +253,17 @@ static void registerDerivative(FunctionDecl* derivedFD, Sema& semaRef) {
+      // derivative is not differentiated again using numerical
+      // differentiation due to unavailable definition.
+      if (auto* CE = dyn_cast<CallExpr>(OverloadedFn))
+        if (auto* FD = CE->getDirectCallee())
```
```suggestion
        if (FunctionDecl* FD = CE->getDirectCallee())
```

vgvassilev commented on tools/ClangPlugin.cpp:
```diff
@@ -407,12 +407,17 @@ namespace clad {
+        // HandleTopLevelDecl can be called recursively in non-standard
+        // setup for code generation.
+        DiffRequest request = m_DiffRequestGraph.getNextToProcessNode();
+        while (request.Function != nullptr) {
```
```suggestion
        while (request.Function) {
```

vgvassilev commented:
@vaithak, is it worth adding the test from #890 here?

vaithak commented:
> @vaithak, is it worth adding the test from #890 here?

Added tests 👍🏼

vgvassilev commented:
I am wondering why the coverage was decreased when we added a test… quite strange. Can you fix the failures and squash into a single commit?

vaithak commented:
> I am wondering why the coverage was decreased when we added a test… quite strange. Can you fix the failures and squash into a single commit?

I need to fix Kokkos test, not sure what is happening there.
For codecov, the lines we added were already covered (although not in the way we wanted). 
I guess adding more lines only increased total lines without changing the total covered lines, hence decreasing the percent by a bit.
This is just a guess though :)

vgvassilev commented:
> > I am wondering why the coverage was decreased when we added a test… quite strange. Can you fix the failures and squash into a single commit?
> 
> I need to fix Kokkos test, not sure what is happening there. For codecov, the lines we added were already covered (although not in the way we wanted). I guess adding more lines only increased total lines without changing the total covered lines, hence decreasing the percent by a bit. This is just a guess though :)

The kokkos failure looks like concurrency failure. We built this file before building clad.so?
